core/utils/watchdog.hpp

```diff
@@ -0,0 +1,157 @@
+/**
+ * Copyright Quadrivium LLC
+ * All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <atomic>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <sstream>
+#include <thread>
+#include <unordered_map>
+
+#include <fmt/format.h>
+#include <boost/asio/io_context.hpp>
+
+#ifdef __APPLE__
+
+#include <mach/mach.h>
+#include <mach/thread_act.h>
+#include <mach/thread_info.h>
+
+inline uint64_t getPlatformThreadId() {
+  thread_identifier_info_data_t info;
+  mach_msg_type_number_t size = THREAD_IDENTIFIER_INFO_COUNT;
+  auto r = thread_info(
+      mach_thread_self(), THREAD_IDENTIFIER_INFO, (thread_info_t)&info, &size);
+  if (r != KERN_SUCCESS) {
+    throw std::logic_error{"thread_info"};
+  }
+  return info.thread_id;
+}
+
+#else
+
+#include <sys/syscall.h>
+#include <unistd.h>
+
+inline uint64_t getPlatformThreadId() {
+  return syscall(SYS_gettid);
+}
+
+#endif
+
+namespace kagome {
+
+  constexpr auto kWatchdogDefaultTimeout = std::chrono::minutes{15};
+
+  class Watchdog {
+   public:
+    using Count = uint32_t;
+    using Atomic = std::atomic<Count>;
+    using Clock = std::chrono::steady_clock;
+    using Timeout = std::chrono::seconds;
+
+    struct Ping {
+      std::shared_ptr<Atomic> count_;
+
+      void operator()() {
+        ++*count_;
+      }
+    };
+
+    void checkLoop(Timeout timeout) {
+      // or `io_context` with timer
+      while (not stopped_) {
+        std::this_thread::sleep_for(std::chrono::seconds{1});
+        check(timeout);
+      }
+    }
+
+    // periodic check
+    void check(Timeout timeout) {
+      std::unique_lock lock{mutex_};
+      auto now = Clock::now();
+      for (auto it = threads_.begin(); it != threads_.end();) {
+        if (it->second.count.unique()) {
+          it = threads_.erase(it);
+        } else {
+          auto count = it->second.count->load();
+          if (it->second.last_count != count) {
+            it->second.last_count = count;
+            it->second.last_time = now;
```

Comment:
It is not atomic check-and-set. Can we check and set at the one moment?

Reply:
No need, `check` uses mutex, and `count = atomic.load()`